subprojects/core/src/main/java/org/gradle/api/services/internal/DefaultBuildServicesRegistry.java

```diff
@@ -137,17 +156,20 @@ public SharedResource forService(Provider<? extends BuildService<?>> service) {
 
     @Override
     public BuildServiceProvider<?, ?> register(String name, Class<? extends BuildService<?>> implementationType, @Nullable BuildServiceParameters parameters, int maxUsages) {
-        if (registrations.findByName(name) != null) {
-            throw new IllegalArgumentException(String.format("Service '%s' has already been registered.", name));
-        }
-        return doRegister(name, uncheckedNonnullCast(implementationType), parameters, maxUsages <= 0 ? null : maxUsages);
+        return withRegistrations(registrations -> {
+            if (registrations.findByName(name) != null) {
+                throw new IllegalArgumentException(String.format("Service '%s' has already been registered.", name));
+            }
+            return doRegister(name, uncheckedNonnullCast(implementationType), parameters, maxUsages <= 0 ? null : maxUsages, registrations);
+        });
     }
 
     private <T extends BuildService<P>, P extends BuildServiceParameters> BuildServiceProvider<T, P> doRegister(
         String name,
         Class<T> implementationType,
         @Nullable P parameters,
-        @Nullable Integer maxParallelUsages
+        Integer maxParallelUsages,
```

Comment:
:x: `maxParallelUsages` can still be `null`, let's keep the `@Nullable` annotation here

Reply:
Good catch! The automatic formatter decided to remove it for some reason.